kbase/narrative
# DATAUP-312 connect to backend - part 1

# Description of PR purpose/changes

First part of connecting the frontend to the backend.
This PR is waaaay bigger than originally intended. Sorry about that. 

The main thrust of this PR is to update the filePathWidget to properly manage its inputs, communicate them to the configure tab for (de)serialization in the cell metadata. 

This PR focuses on getting the bulkImportCell metadata set up right and refactoring the filePathWidget to handle its inputs better. Next will give a similar treatment to the paramsWidget. That'll be followed with some input validation and Python code setup, and finally calling a to-be-written kernel-side function to start the bulk job. I think this one's the heaviest lift, as it does the most setup for the rest.

There's a few core parts I'll outline here, then follow up through the code changes.

**Change the metadata.**  
The app cell and others holds inputs in the cell metadata. This lets us reload them on page load or on tab switching. The bulk import cell does the same, but we have to manage some differences. First, we have multiple apps that we're wrangling, so the inputs have to be set by app id. Second, there's a split between file path parameters (mapping to the staging area files and the output objects) and other parameters. The file path parameters need to be stored as an ordered array of objects, and we need to know what those ids are so we can set up the filePathWidget efficiently. So the structure that hold this data has been changed from a fairly flat mapping of app id -> set of parameters for that app to:
```
{
    appId: {
        filePaths: [{
            file1: value1, 
            file2: value2, 
            ...
        }],
        params: {
            param1: value,
            ....
        }
    }
}
```

There's other metadata changes, too. The structure's described in detail mostly in the code, but I'll write it up in some documentation as well. Mainly, this adds fields for `fileParamIds` and `otherParamIds` under the `app` heading.

**Change the file path widget**
Lots of setup for the file path widget (FPW) has changed. It now keeps an internal data model (kind of a view model, I think), for what data it should be displaying. This gets updated from the user inputs. So, now, instead of user inputs directly being pushed up the paramsBus to the configure widget, they get trapped and held internally to the filePathWidget, which then sends the entire change at once up for storage. This means changes to a few things.
1. Update how the FPW gets instantiated. Now it knows what default parameters should be, and what the previous layout was, so it sets itself back up as it used to be.
2. Update what happens on row addition / deletion. Previously, these were pretty lightweight operations that just plopped down some new widgets or removed them. Now, it properly stops widgets and removes them from the DOM before deleting the row, and modifies the data model to keep up. This is all pretty similar to how `appWidgets2/input/sequenceInput` works. They could probably be combined at some point, but I don't think that's worth the effort yet.
3. Update how bus messages get communicated. There were a few superfluous ones that, I think, were copy-pasted from the older paramsWidget. Also, others were blindly passed up to the configure widget without either directing to the right channel or managing the right value. This trims down and focuses (maybe too narrowly) which messages are meaningful.

There's some other random code cleanup and tuning as well here that seemed necessary.

# Jira Ticket / Issue #
https://kbase-jira.atlassian.net/browse/DATAUP-312

# Testing Instructions
* Details for how to test the PR:
* First, you'll need a new Bulk Import cell as the expectations from metadata have changed.
* The file path area should now auto-populate with inputs of various files.
* Any changes should get stored in a way that:
  * gets restored on page reload (after saving the narrative)
  * gets restored on swapping between file input types using the left-column-nav
  * gets restored on tab swapping
  * (this doesn't affect the non-file params yet)

# Dev Checklist:


# Updating Version and Release Notes (if applicable)
(none yet)
